internal/blobinfocache/types.go

```diff
@@ -0,0 +1,63 @@
+package blobinfocache
+
+import (
+	"github.com/containers/image/v5/types"
+	digest "github.com/opencontainers/go-digest"
+)
+
+const (
+	// Uncompressed is the value we store in a blob info cache to indicate that we know that the blob in
+	// the corresponding location is not compressed.
+	Uncompressed = "uncompressed"
+	// UnknownCompression s the value we store in a blob info cache to indicate that we don't know if the
+	// blob in the corresponding location is compressed or not.
+	UnknownCompression = "unknown"
+)
+
+// BICReplacementCandidate2 is an item returned by BlobInfoCache2.CandidateLocations2.
+type BICReplacementCandidate2 struct {
+	Digest         digest.Digest
+	CompressorName string // either the Name() of a known pkg/compression.Algorithm, or Uncompressed or UnknownCompression
+	Location       types.BICLocationReference
+}
+
+// BlobInfoCache2 extends BlobInfoCache by adding the ability to track information about what kind
+// of compression was applied to the blobs it keeps information about.
+type BlobInfoCache2 interface {
+	types.BlobInfoCache
+	blobInfoCache2Additions
+}
+
+// blobInfoCacheCommon contains methods which are common to both BlobInfoCache and BlobInfoCache2:
+// methods which store and retrieve mappings of uncompressed and possibly-compressed digests.
+type blobInfoCacheCommon interface {
+	UncompressedDigest(anyDigest digest.Digest) digest.Digest
+	RecordDigestUncompressedPair(anyDigest digest.Digest, uncompressed digest.Digest)
+}
+
+// BlobInfoCache2WithoutV1 contains extended versions of the BlobInfoCache methods which add the
+// ability to track information about what kind of compression was applied to the blobs it keeps
+// information about, and the same methods for managing mappings of digests.
+type BlobInfoCache2WithoutV1 interface {
+	blobInfoCacheCommon
+	blobInfoCache2Additions
+}
+
+// blobInfoCache2Additions contains extended versions of the BlobInfoCache methods which add the
+// ability to track information about what kind of compression was applied to the blobs it keeps
+// information about.
+type blobInfoCache2Additions interface {
+	// RecordKnownLocation2 records that a blob with the specified digest exists within the
+	// specified (transport, scope) scope, compressed with the specified compression algorithm
+	// (one recognized by pkg/compression.AlgorithmByName()), or Uncompressed or with
+	// UnknownCompression, and can be reused given the opaque location data.
+	RecordKnownLocation2(transport types.ImageTransport, scope types.BICTransportScope, digest digest.Digest, compressorName string, location types.BICLocationReference)
```

Comment:
(Looking at the docker loop) the compressor name is really a property of the digest, not of any specific location; so this could probably make sense as a separate “record compression for digest” API.

Or maybe the API should be a single call to have only a single BoltDB lock+use, but internally be recorded in two different ways? (Similarly, `CandidateLocations2` might still return both together.)

Recording the compression per-digest instead of per-location would also directly lead to `copyBlobFromStream`, which is _already_ detecting compression, being in charge of recording that — _both_ for the source and the destination, instead of `dockerImage{Source.GetBlob,Destination,PutBlob}`.

Reply:
I'm starting to rough out a `RecordDigestCompressorName()` method, which for boltdb stores compressor names (or "uncompressed" or "unknown") in a top-level "digestCompressor" bucket keyed by blob digest, to see how well that goes.